design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,260 @@
+and supported, something that does not exist with minor releases.
+
+#### The Deprecation Flow (for Minor releases)
+As stated before, only features that have not yet been graduated
+can be discontinued between minor releases.
+Therefore, a feature can be deprecated in Alpha and Beta stages only.
+
+These are the steps needed to deprecate a feature:
+- Proposal: Prepare a proposal to discontinue a feature with proper
+  reasoning, phases, exact timelines and functional alternatives (if any).
+  The proposal should be reviewed and approved.
+- Notification: Notify the project community about the feature
+  discontinuation based on the approved proposal.
+  All details of the plan should be provided to allow users and possibly
+  d/s projects to adjust.
+  Use all community media options to distribute this information
+  (e.g. mailing list, slack channel, community meetings).
+- Deprecation warnings: Add deprecation warnings at runtime to warn users
+  that the feature is planned to be discontinued.
+  Warnings should be raised when:
+  - Feature API fields are accessed.
+  - Feature FG is activated.
+  - Behavior related to the feature is detected (optional).
+- Removal: Feature removal involves dropping the core functionality
+  of a feature and its exposed API.
+  - The core implementation can be dropped in two steps:
+    - The FG is dropped by assuring it is never reported as set
+      (i.e. even if it is left by the operator configured, internally
+       it is ignored).
+      At this stage, the core implementation will follow the FG conditions
+      and therefore from the outside the feature is inactive.
+    - In case there are no side effects, the core implementation code can
+      be removed.
+  - The API types are not to be removed, as it may have implications
+    with the underlying storage which has already persisted them.
+    Kubernetes has not removed fields, it just kept them around with the
+    warning that they have been deprecated and no longer available.
+
+    While keeping fields around for a period of a release or two makes
+    sense, beyond a limited period it adds a burden on dragging leftover
+    fields around to eternity.
+
+> **Note**: The only reference seen on why fields should not be removed
+> was mentioned [here](https://github.com/kubernetes/kubernetes/issues/52185).
+> But it is unclear if this is relevant for Alpha stage features.
+
+### Exceptions
+While the project strives to maintain a stable contract with its users,
+there may be scenarios where the policy describe here will not be a fit.
+
+Therefore, it should be acceptable to have exceptions from time to time
+given a very good reasoning and an agreement from 2/3 of the project
+maintainers.
+
+## Implementation Phases
+N/A
+
+## Open Issues
+1. Removal of fields from the CRD: Is it relevant for Alpha
+   and Beta features?
+2. The details on how exactly to mark an Alpha/Beta field.
```

Comment:
Please please include it.

Reply:
My opinion on this item changes each time I think about it :)

At the moment, this proposal does not enter into the details of how to implement the process & policy.
This has the advantage of agreeing on the WHAT and leaving the HOW to a follow up.

I have a feeling there are many details that may pop up when we'll look into the implementation details. E.g. what data to add to the design proposal template, what information & format to mark on fields, how to remove a FG in the code, etc.

Would you be ok with us just adding the need to add this marking, mentioning only what is needed there (e.g. the version of the feature and the version of the release it was introduced at)?

Reply:
Yes.
Agreed on taking this async.